runtime/common/src/precompile/homa.rs

```diff
@@ -0,0 +1,442 @@
+#[repr(u32)]
+pub enum Action {
+	Mint = "mint(address,uint256)",
+	RequestRedeem = "requestRedeem(address,uint256,bool)",
+	GetExchangeRate = "getExchangeRate()",
+	GetEstimatedRewardRate = "getEstimatedRewardRate()",
+	GetCommissionRate = "getCommissionRate()",
+	GetFastMatchFee = "getFastMatchFee()",
+}
+
+impl<Runtime> Precompile for HomaPrecompile<Runtime>
+where
+	Runtime: module_evm::Config + module_homa::Config + module_prices::Config,
+	module_homa::Pallet<Runtime>: HomaManager<Runtime::AccountId, Balance>,
+{
+	fn execute(input: &[u8], target_gas: Option<u64>, _context: &Context, _is_static: bool) -> PrecompileResult {
+		let input = Input::<
+			Action,
+			Runtime::AccountId,
+			<Runtime as module_evm::Config>::AddressMapping,
+			Runtime::Erc20InfoMapping,
+		>::new(input, target_gas_limit(target_gas));
+
+		let gas_cost = Pricer::<Runtime>::cost(&input)?;
+
+		if let Some(gas_limit) = target_gas {
+			if gas_limit < gas_cost {
+				return Err(PrecompileFailure::Error {
+					exit_status: ExitError::OutOfGas,
+				});
+			}
+		}
+
+		let action = input.action()?;
+
+		match action {
+			Action::Mint => {
+				let who = input.account_id_at(1)?;
+				let amount = input.balance_at(2)?;
+
+				log::debug!(
+					target: "evm",
+					"homa: mint, who: {:?}, amount: {:?}",
+					&who, amount
+				);
+
+				<module_homa::Pallet<Runtime> as HomaManager<Runtime::AccountId, Balance>>::mint(who, amount).map_err(
+					|e| PrecompileFailure::Revert {
+						exit_status: ExitRevert::Reverted,
+						output: Into::<&str>::into(e).as_bytes().to_vec(),
+						cost: target_gas_limit(target_gas).unwrap_or_default(),
+					},
+				)?;
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: vec![],
+					logs: Default::default(),
+				})
+			}
+			Action::RequestRedeem => {
+				let who = input.account_id_at(1)?;
+				let amount = input.balance_at(2)?;
+				let fast_match = input.bool_at(3)?;
+
+				log::debug!(
+					target: "evm",
+					"homa: request_redeem, who: {:?}, amount: {:?}, fast_match: {:?}",
+					&who, amount, fast_match
+				);
+
+				<module_homa::Pallet<Runtime> as HomaManager<Runtime::AccountId, Balance>>::request_redeem(
+					who, amount, fast_match,
+				)
+				.map_err(|e| PrecompileFailure::Revert {
+					exit_status: ExitRevert::Reverted,
+					output: Into::<&str>::into(e).as_bytes().to_vec(),
+					cost: target_gas_limit(target_gas).unwrap_or_default(),
+				})?;
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: vec![],
+					logs: Default::default(),
+				})
+			}
+			Action::GetExchangeRate => {
+				let rate =
+					<module_homa::Pallet<Runtime> as HomaManager<Runtime::AccountId, Balance>>::get_exchange_rate();
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::default().encode_u128(rate.into_inner()),
```

Comment:
Not sure about what fixed point arithmetic libraries are best in solidity, but I figure that this format will be usable enough

Reply:
I think it's ok. The same as here.
https://github.com/AcalaNetwork/Acala/blob/45c41201fc1dc7702d5cb27eb214f0c2996c4324/runtime/common/src/precompile/oracle.rs#L102